packages/admin-ui-hooks/src/use-intersection-observer.ts

```diff
@@ -0,0 +1,47 @@
+import { useEffect, useRef, useState } from 'react'
+
+/**
+ * https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API
+ */
+export function useIntersectionObserver({
+  // Defaults to the browser viewport
+  root = null,
+  rootMargin = undefined,
+  threshold = 0,
+  callback,
+}: IntersectionObserverInit & {
+  callback: (entry: IntersectionObserverEntry) => void
+}) {
+  const [node, setNode] = useState<HTMLDivElement>()
+  const observer = useRef<IntersectionObserver>()
+
+  useEffect(() => {
+    observer?.current && observer.current.disconnect()
+
+    observer.current = new IntersectionObserver(
+      ([entry]) => {
+        callback(entry)
+      },
+      {
+        root,
+        rootMargin,
+        threshold,
+      }
+    )
+
+    const { current: currentObserver } = observer
+
+    // We need to disconnect in case the component using
+    // this hook changes the node that the observer is
+    // observing, otherwise we'll polute the DOM with
+    // unused observables. This ensures that the observer
+    // is acting only over the node we care about.
+    currentObserver.disconnect()
```

Comment:
When the useEffect dependencies change the cleanup function is called, so every time the node dependency changes, before run the useEffect, the previous cleanup function is called. Every time the node changes the `line 43` is called.

Reply:
https://github.com/vtex/admin-ui/pull/267#discussion_r913135653